command/src/main/java/discord4j/command/ProviderContext.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * This file is part of Discord4J.
+ *
+ * Discord4J is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Discord4J is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Discord4J.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package discord4j.command;
+
+import javax.annotation.Nullable;
+import java.util.Objects;
+import java.util.Optional;
+
+/** Context from a {@link CommandProvider} for a command's execution. */
+public final class ProviderContext {
+
+    /**
+     * Constructs a {@code ProviderContext}.
+     *
+     * @param command The command for possible execution.
+     * @param context The context for the command's execution. May be null.
+     * @return A constructed {@code ProviderContext}.
+     */
+    public static ProviderContext of(final Command command, @Nullable final Object context) {
```

Comment:
I wonder if its possible to make contexts type safe. Since providers control their own command impls, we could make the context type generic and/or provide the expected Class of the context. I just feel a little uncomfortable everytime unchecked casts are required in a public api

Reply:
If we only did a generic then this restricts Providers to only 1 type of Context they can utilize (since Command implementations can only expect one type to be provided by the Provider), but, this is relatively easy to work with (just change Command to Command<T>).
We could require a Class and Object, and change execute in Command to `<T> Mono<Void> execute(MessageCreateEvent, Class<T>, T)`
Either is fine with me

Reply:
I like that option